### ISSUE TITLE
Proposal - Support content redirects for blob downloads

### ISSUE THREAD

**Issue by BrianBland** [docker/distribution#49]
To reduce the amount of downstream bandwidth congestion for the registry, we should optionally support content-redirects to the underlying storage system for blob download requests. For instance, when using S3 for object storage, we can redirect the client to a temporary signed S3 download URL, removing the need to transfer data through the registry itself.

Because a download URL is strongly tied to where the file itself is stored, we need to add a method to the `storagedriver.StorageDriver` interface which can return a static URL path for a given blob path. I'm proposing we add this in the form of `URLFor(path string) (string, error)`, which will return a static download URL which is expected to be valid for at least a short period of time. Because not all storage systems (local filesystem as a basic example) do not support URL referencing, we must also add a new error type: `UnsupportedMethodErr`, which may also be used for future optional extensions.

This unfortunately gets a bit hairy when we consider alternative static URL providers (CDNs), such as CloudFront, which in particular generates signed URLs based on the path of the hosted content. For s3+cloudfront, the CloudFront URL provider would need to know the base path at which the files are stored in the S3 bucket, which is a configuration parameter of the S3 storagedriver itself.

One solution here is to add an optional LayerHandler subsystem, which can take a `storage.Layer` and serve its contents via a `Resolve(layer storage.Layer) (http.Handler, error)` method. In the case of s3+cloudfront, the CloudFront LayerHandler constructs the S3 object URL and translates it to a CloudFront URL then serves a 307 Temporary Redirect with a Location header referencing the new URL. An example translation would look like `https://mybucket.s3.amazonaws.com/myobject?AWSAccessKeyId=ACCESSKEY&Expires=EXPIRATION&Signature=S3SIGNATURE` -> `https://mycloudfrontsubdomain.cloudfront.net/myobject?Expires=EXPIRATION&Signature=CFSIGNATURE&Key-Pair-Id=CFKEYPAIRID`, in which the CloudFront URL generator utilizes only the path from the S3 URL (or any URL, this isn't specific to S3). The LayerHandler would be configured with its own optional name/parameters map, similar to to the storagedriver and auth systems. When not provided, no alternate Layer serving will be performed.

There is one addition to the configuration struct required for this change:
* Add a field to the `HTTP` section for `LayerHandler`.
  * Valid options are `delegate` (use the underlying storagedriver with no translation) and `cloudfront` for now. If either of these are specified, the registry will use URL redirects for serving blobs.
    * For `cloudfront`, the following parameters must be supplied: `base`, `keyid`, and `keysecret`, just as in the python registry.

tl;dr Making the following additions:

```go
// New storagedriver error type
// UnsupportedMethodErr may be returned in the case where a StorageDriver implementation does not support an optional method.
var UnsupportedMethodErr = errors.New("Unsupported method")

type StorageDriver interface {
    // New method
    // URLFor returns a URL which may be used to retrieve the content stored at the given path.
    // May return an UnsupportedMethodErr in certain StorageDriver implementations.
    URLFor(path string) (string, error)
}

// New interface
type LayerHandler interface {
    // Resolve returns an http.Handler which can be used to serve the contents of the given storage.Layer.
    Resolve(layer storage.Layer) (http.Handler, error)
}
```

This should solve #37 and #38

**Comment by BrianBland** [comment]
@stevvooe @dmp42 @AndreyKostov

**Comment by stevvooe** [comment]
@BrianBland This looks good. 

A few things:

1. Could you enumerate the proposed interfaces and modifications?

   For example:
   
   ```
   type URLProvider interface {
       // URLFor yada yada
       URLFor(...) ...
   }
   
   type StorageDriver interface {
       MyNewMethod(...) ...
   }
   ```

2. The user of the blob redirector shouldn't really have knowledge of backend paths. Maybe it should be something along these lines:

   ```
   type LayerRedirector interface {
       Redirect(layer storage.Layer) (http.Handler, error)
   }
   ```
   
   The path mapping can then be done on the backend.

**Comment by BrianBland** [comment]
1. Sure, I'll update the proposal with cleaner formatting.
2. The redirector often needs more than just the layer name/digest/metadata, as it is likely dependent on the full static URL path of the object on the underlying storage system. This could work with the `storage.Layer` interface if we decide to add a `ContentURL() (string, error)` method to this interface. Thoughts?

**Comment by BrianBland** [comment]
Updated, but I really need to come up with a better name than `LayerDispatcher`...

**Comment by AndreyKostov** [comment]
We probably want a way to define which methods are allowed to remain unimplemented, because it remains rather vague this way.

**Comment by stevvooe** [comment]
@BrianBland How about this:

```
type LayerHandler interface {
    Resolve(layer storage.Layer) (http.Handler, error)
}
```

**Comment by BrianBland** [comment]
@stevvooe sounds reasonable enough.

There's one more thing that came up in a separate discussion. Should we support an expiration date argument to `URLFor()`? This could be respected by the S3 storagedriver, but other storagedrivers may ignore this. Additionally, the `LayerHandler` would have to be configured independently with an expiration time for redirect URLs if relevant, so this may end up being too misleading.

**Comment by stevvooe** [comment]
@BrianBland What if it took an opaque parameter map for configuration of the URL? We may request an "expiry" now, but in the future, maybe there is geo region or some crypto parameter. Something like this seems futureproof:

```
URLFor(path string, options map[string]interface{}) (string, error)
```

**Comment by stevvooe** [comment]
Satisfied by #55.
